### PR TITLE
Add a --remove-untracked option to the install command.

### LINKED ISSUE #648: How do I uninstall packages that have been removed from pyproject.toml?
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://poetry.eustace.io/docs/) and believe that my question is not covered.

## Question

pipenv has `pipenv clean` which removes all packages that are no longer in Pipfile.lock from the virtualenv. What is the equivalent with poetry? The docs say there is no need for `pipenv sync` and `pipenv clean` due to the `remove` command https://github.com/sdispater/poetry#remove-command

But how does this work? I would need to run `poetry remove XXX` against all my virtualenvs - manually keeping track of all packages that have been removed and remembering to remove them myself?

Is there not any way to sync my poetry.lock to my virtualenv to ensure the virtualenv only has the packages specified in poetry.lock installed?

I have a virtualenv on each box I deploy to, and keep pyproject.toml and poetry.lock under source code control. I then want a way to ensure each of those virtualenvs only has the packages in poetry.lock installed.

### PR DESCRIPTION
**Pull request by PetterS**
Fixes #648. 

As discussed in the linked issue. @finswimmer seemed to approve this flag, @sdispater. 

This is a feature that we are missing since moving away from pipenv. Removing old packages is part of reproducibility.

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

### CODE REVIEW

**Comment by PetterS** [comment]
Could anyone have a look at this? I am happy to finish this up with tests and documentation.

**Comment by Mattwmaster58** [comment]
I'm not a poetry team member, but I took a quick look at this anyway, and I have to say you did a thorough job as far as I can tell 👍

**Comment by PetterS** [comment]
Thanks Matt!

@stephsamson, do you think you could take a look?

**Comment by ulgens** [comment]
How does this handle packages like pip and setuptools which are not in lock file but should be kept in venv?

**Comment by PetterS** [comment]
> How does this handle packages like pip and setuptools which are not in lock file but should be kept in venv?

This is a great question and the answer is that, unfortunatley, currently it does not. This needs to be addressed.

**Comment by PetterS** [comment]
@sdispater did you have a chance to look at this? This is functionality that we really would like to have in poetry (we used `pipenv clean` before).

Thanks,
Petter

**Comment by PetterS** [comment]
Is there something I could do here, besides waiting for a review? We'd love to have this in Poetry!

**Comment by ayroblu** [comment]
Bumping this cause we're waiting on this

**Comment by PetterS** [comment]
> Bumping this cause we're super interested in this

Yes, the lack of this flag has been a pain point in my team since we switched from pipenv to poetry. People are `rm -rf` their venvs and installing from scratch.

How can we proceed here? @stephsamson @sdispater @finswimmer

**Comment by PetterS** [comment]
@stephsamson Done. Moved the logic to the solver.

File: poetry/puzzle/solver.py [hunk 1]
```diff
@@ -21,14 +21,15 @@
 
 
 class Solver:
-    def __init__(self, package, pool, installed, locked, io):
+    def __init__(self, package, pool, installed, locked, io, remove_untracked=False):
```
Comment (stephsamson) [inline poetry/puzzle/solver.py:24]: A very minor change needed as we make a concentrated effort to clean up the Poetry codebase by adding type annotations whenever we touch a piece of code. Thank you! 🙌 

```suggestion
    def __init__(
        self,
        package,  # type: ProjectPackage
        pool,  # type: Pool
        installed,  # type: Repository
        locked,  # type: Repository
        io,  # type: ConsoleIO
        remove_untracked=False  # type: bool
    ):
```

Comment (PetterS) [reply poetry/puzzle/solver.py]: Thanks. I really like type annotations.

**Comment by PetterS** [comment]
OK, those type annotations did not work. I'll try to fix them.

File: poetry/puzzle/solver.py (hunk 1 above)
Comment (snejus) [reply poetry/puzzle/solver.py]: Is there a reason behind adding them in the comments rather than directly?

**Pull request merged**